File: app/src/main/java/com/wolcano/musicplayer/music/ui/adapter/AlbumAdapter.java
package com.wolcano.musicplayer.music.ui.adapter;

import android.app.Activity;
import androidx.annotation.NonNull;
import androidx.appcompat.widget.PopupMenu;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import android.view.ContextThemeWrapper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import com.squareup.picasso.Picasso;
import com.wolcano.musicplayer.music.R;
import com.wolcano.musicplayer.music.databinding.ItemAlbumBinding;
import com.wolcano.musicplayer.music.mvp.listener.Bind;
import com.wolcano.musicplayer.music.mvp.models.Album;
import com.wolcano.musicplayer.music.ui.dialog.Dialogs;
import com.wolcano.musicplayer.music.utils.Utils;
import java.util.List;

public class AlbumAdapter extends RecyclerView.Adapter<AlbumAdapter.ViewHolder> {

    private List<Album> arraylist;
    private Activity context;

    public AlbumAdapter(Activity context, List<Album> arraylist) {
        this.context = context;
        this.arraylist = arraylist;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemAlbumBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()), R.layout.item_album, parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {

        holder.binding.setAlbum(arraylist.get(position));
        holder.binding.executePendingBindings();
        Album album = holder.binding.getAlbum();
        holder.binding.line1.setText(album.getName());
        holder.binding.line2.setText(album.getArtist());
        String albumUri = "content://media/external/audio/albumart/" + album.getId();
        Picasso.get().load(albumUri).placeholder(R.drawable.album_art).into(holder.binding.albumArt);
        setOnPopupMenuListener(holder, position);

    }
    @Override
    public int getItemCount() {
        return (null != arraylist ? arraylist.size() : 0);
    }
    private void setOnPopupMenuListener(AlbumAdapter.ViewHolder holder, final int position) {
        holder.binding.more.setOnClickListener(v -> {
            try {
                ContextThemeWrapper contextThemeWrapper = new ContextThemeWrapper(v.getContext(), R.style.PopupMenuToolbar);

                PopupMenu popup = new PopupMenu(contextThemeWrapper, v);
                popup.getMenuInflater().inflate(R.menu.popup_menu_libary_albums, popup.getMenu());
                popup.show();
                popup.setOnMenuItemClickListener(item -> {

                    switch (item.getItemId()) {
                        case R.id.copy_to_clipboard:
                            Dialogs.copyDialog(context, arraylist.get(position).getName(),arraylist.get(position).getArtist());
                            break;
                        default:
                            break;
                    }

                    return true;
                });
            } catch (Exception e) {
                e.printStackTrace();
            }
        });
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private ItemAlbumBinding binding;

        public ViewHolder(ItemAlbumBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
            this.binding.getRoot().setOnClickListener(this::onClick);
        }

        @Override
        public void onClick(View view) {
            try {

                Album album = arraylist.get(getAdapterPosition());
                Utils.navigateToAlbum(context, album.getId(),
                        album.getName());


            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }
}

File: app/src/main/java/com/wolcano/musicplayer/music/ui/adapter/AlbumAdapter.kt
package com.wolcano.musicplayer.music.ui.adapter

import android.app.Activity
import android.view.*
import androidx.appcompat.widget.PopupMenu
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso
import com.wolcano.musicplayer.music.R
import com.wolcano.musicplayer.music.data.model.Album
import com.wolcano.musicplayer.music.databinding.ItemAlbumBinding
import com.wolcano.musicplayer.music.ui.dialog.Dialogs.copyDialog
import com.wolcano.musicplayer.music.utils.Utils.navigateToAlbum

class AlbumAdapter(private val context: Activity, private val albumList: List<Album>) :
    RecyclerView.Adapter<AlbumAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding: ItemAlbumBinding = DataBindingUtil.inflate(
            LayoutInflater.from(parent.context),
            R.layout.item_album,
            parent,
            false
        )
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.binding.album = albumList[position]
        holder.binding.executePendingBindings()

        val album = holder.binding.album

        holder.binding.line1.text = album?.name
        holder.binding.line2.text = album?.artist
        val albumUri = "content://media/external/audio/albumart/" + album?.id
        Picasso.get().load(albumUri).placeholder(R.drawable.album_art).into(holder.binding.albumArt)
        setOnPopupMenuListener(holder, position)
    }

    override fun getItemCount(): Int {
        return albumList.size
    }

    private fun setOnPopupMenuListener(holder: ViewHolder, position: Int) {
        holder.binding.more.setOnClickListener { v ->
            try {
                val contextThemeWrapper =
                    ContextThemeWrapper(v.context, R.style.PopupMenuToolbar)
                val popup =
                    PopupMenu(contextThemeWrapper, v)
                popup.menuInflater.inflate(R.menu.popup_menu_libary_albums, popup.menu)
                popup.show()
                popup.setOnMenuItemClickListener { item: MenuItem ->
                    if (item.itemId == R.id.copy_to_clipboard) {
                        copyDialog(
                            context,
                            albumList[position].name,
                            albumList[position].artist
                        )
                    }
                    true
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    inner class ViewHolder(val binding: ItemAlbumBinding) : RecyclerView.ViewHolder(binding.root),
        View.OnClickListener {
        override fun onClick(view: View) {
            try {
                val album: Album = albumList[absoluteAdapterPosition]
                navigateToAlbum(
                    context, album.id,
                    album.name
                )

            } catch (e: java.lang.Exception) {
                e.printStackTrace()
            }
        }

        init {
            this.binding.root.setOnClickListener(::onClick)
        }
    }

}